### PR TITLE
feat(ingest): dbt column-level meta mappings + `add_terms` operation

### PR DESCRIPTION
**Pull request by hsheth2**
## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/dbt.py [hunk 1]
```diff
@@ -1175,11 +1091,9 @@ def loadManifestAndCatalog(
         else:
             sources_results = {}
 
-        manifest_schema = dbt_manifest_json.get("metadata", {}).get(
-            "dbt_schema_version"
-        )
-        manifest_version = dbt_manifest_json.get("metadata", {}).get("dbt_version")
-        manifest_adapter = dbt_manifest_json.get("metadata", {}).get("adapter_type")
+        manifest_schema = dbt_manifest_json["metadata"].get("dbt_schema_version")
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/dbt.py:1094]: Will this throw a keyError (will `metadata` key be always present)?

File: metadata-ingestion/src/datahub/ingestion/source/dbt.py [hunk 2]
```diff
@@ -1422,40 +1329,58 @@ def get_workunits(self) -> Iterable[MetadataWorkUnit]:
...
             )
 
         yield from self.stale_entity_removal_handler.gen_removed_entity_workunits()
 
+    def filter_nodes(self, all_nodes: List[DBTNode]) -> List[DBTNode]:
+        nodes = []
+        for node in all_nodes:
+            key = node.dbt_name
+
+            if not self.config.node_type_pattern.allowed(node.node_type):
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/dbt.py:1366]: Do these need to be added to the report as filtered?

File: metadata-ingestion/src/datahub/utilities/mapping.py [hunk 3]
```diff
@@ -102,12 +104,21 @@ def process(self, raw_props: Dict[str, Any]) -> Dict[str, Any]:
...
+
                     if operation:
-                        if isinstance(operation, str):
+                        if isinstance(operation, str) or isinstance(operation, list):
                             operations_value_set = operations_map.get(
                                 operation_type, set()
                             )
-                            operations_value_set.add(operation)  # type: ignore
+                            if isinstance(operation, list):
+                                breakpoint()
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/utilities/mapping.py:118]: Remove `breakpoint()` call :-)

File: metadata-ingestion/src/datahub/utilities/mapping.py [hunk 4]
```diff
@@ -27,6 +28,7 @@ class Constants:
     GROUP_OWNER = "group"
     OPERAND_DATATYPE_SUPPORTED = [int, bool, str, float]
     TAG_PARTITION_KEY = "PARTITION_KEY"
+    SEPARATOR = "separator"
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/utilities/mapping.py:31]: Is the value `separator` intentional?

File: metadata-ingestion/src/datahub/ingestion/source/dbt.py [hunk 5]
```diff
@@ -434,32 +442,52 @@ class DBTNode:
     tags: List[str] = field(default_factory=list)
     compiled_sql: Optional[str] = None
 
-    def __repr__(self):
-        fields = tuple("{}={}".format(k, v) for k, v in self.__dict__.items())
-        return self.__class__.__name__ + str(tuple(sorted(fields))).replace("'", "")
+    def get_db_fqn(self) -> str:
+        if self.database is not None:
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/dbt.py:446]:
```suggestion
        if self.database:
```

File: metadata-ingestion/src/datahub/ingestion/source/dbt.py [hunk 6]
```diff
@@ -434,32 +442,52 @@ class DBTNode:
...
     catalog_node: dict, manifest_node: dict, tag_prefix: str
 ) -> List[DBTColumn]:
     columns = []
 
+    catalog_columns = catalog_node["columns"]
     manifest_columns = manifest_node.get("columns", {})
 
-    raw_columns = catalog_node["columns"]
+    for key in catalog_columns:
+        catalog_column = catalog_columns[key]
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/dbt.py:476]:
```suggestion
    for key, catalog_column in catalog_columns.items():
```

File: metadata-ingestion/src/datahub/utilities/mapping.py [hunk 7]
```diff
@@ -102,12 +104,21 @@ def process(self, raw_props: Dict[str, Any]) -> Dict[str, Any]:
...
-                        if isinstance(operation, str):
+                        if isinstance(operation, str) or isinstance(operation, list):
                             operations_value_set = operations_map.get(
                                 operation_type, set()
                             )
-                            operations_value_set.add(operation)  # type: ignore
+                            if isinstance(operation, list):
+                                breakpoint()
+                                operations_value_set.update(operation)  # type: ignore
+                            else:
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/utilities/mapping.py:120]: check or assert that it is a list for future proofing?

File: metadata-ingestion/src/datahub/utilities/mapping.py (hunk 4 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/utilities/mapping.py:31]: yes

File: metadata-ingestion/src/datahub/ingestion/source/dbt.py (hunk 1 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/source/dbt.py:1094]: yes it should always be present - I checked the dbt json schemas

File: metadata-ingestion/src/datahub/utilities/mapping.py (hunk 7 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/utilities/mapping.py:120]: the if statements above should check this

File: metadata-ingestion/src/datahub/ingestion/source/dbt.py (hunk 2 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/source/dbt.py:1365]: the dbt report doesn't have a filtered object, so I'm just leaving that behavior as-is (this was just a refactor)

we'll add it to the backlog

**Review by rslanka — APPROVED** [review]
LGTM!

File: metadata-ingestion/src/datahub/utilities/mapping.py [hunk 8]
```diff
@@ -102,12 +104,20 @@ def process(self, raw_props: Dict[str, Any]) -> Dict[str, Any]:
...
                         operation_key, operation_type, operation_config, maybe_match
                     )
+                    if operation_type == Constants.ADD_TERMS_OPERATION:
+                        # add_terms operation is a special case where the operation value is a list of terms.
+                        # We want to aggregate these values with the add_term operation.
+                        operation_type = Constants.ADD_TERM_OPERATION
+
                     if operation:
-                        if isinstance(operation, str):
+                        if isinstance(operation, str) or isinstance(operation, list):
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/utilities/mapping.py:113]:
```suggestion
                        if isinstance(operation, (str, list)):
```

Comment (hsheth2) [reply metadata-ingestion/src/datahub/utilities/mapping.py]: done

**Pull request merged**